### PR TITLE
fix: unhandled exception error catch

### LINKED ISSUE #1972: Uncaught exception with concurrent queries and inline begin
The client throws an unhandled exception if the initial queries in the transaction are called concurrently and one of the queries fails(except for the first one). You can see the example below. The issue is resolved if `transaction.begin()` is called explicitly before the concurrent queries. 

#### Environment details

  - OS: macOS 14.2 (23C64)
  - Node.js version: v20.1.0
  - npm version: 9.6.4
  - `@google-cloud/spanner` version: reproducible on master 

#### Steps to reproduce

Minimal repro

```ts
// We discovered this issue with ABORTED errors, but
// it's easier to reproduce with this non-existent table query.
const nonExistantTableQuery = 'SELECT * FROM nonExistingTable';
await database.runTransactionAsync(async tx => {
  try {
    await Promise.all([
      tx!.run(selectSql),
      tx!.run(nonExistantTableQuery),
    ]);
    await tx.commit();
  } catch (err) {
    await tx.rollback();
  }
});

```

This code will result in an uncaught exception and it will crash node process.

I think the issue is a missing error handler in this code. 

https://github.com/googleapis/nodejs-spanner/blob/6c0eef16d85e6da4993b246337483a818a4521b5/src/transaction.ts#L1336-L1338

Adding this handler fixes the issue
```TS
.on('error', err => streamProxy.emit('error', err))
```

Happy to open PR with test if this seems correct

### PR DESCRIPTION
**Pull request by alkatrivedi**
fixes: #1972

### CODE REVIEW

File: test/spanner.ts [hunk 1]
```diff
@@ -3211,6 +3211,21 @@ describe('Spanner with mock server', () => {
...
+    it('should catch exception error during invalid queries while using inline begin transaction', async () => {
+      const database = newTestDatabase();
+      try {
+        await database.runTransactionAsync(async tx => {
+          await tx!.run(selectSql);
+          await tx!.run(invalidSql);
+          await tx.commit();
+        });
+      } catch (err) {
+        (err as grpc.ServiceError).message.includes('Table FOO not found');
```
Comment (surbhigarg92) [inline test/spanner.ts:3223]: Need an assert here.

Comment (alkatrivedi) [reply test/spanner.ts]: Done

**Review by surbhigarg92 — APPROVED** [review]

File: test/spanner.ts [hunk 2]
```diff
@@ -1314,6 +1314,43 @@ describe('Spanner with mock server', () => {
...
+          assert.ok(
+            requests[1].transaction!.id,
+            'Transaction ID is not used for retries.'
+          );
+          assert.ok(
+            requests[1].resumeToken,
+            'Resume token is not set for the retried'
+          );
+        });
+
```
Comment (olavloite) [inline test/spanner.ts:1353]: Can we add a test that verifies the behavior if the initial `ExecuteStreamingSql` call succeeds, but the stream then fails with `UNAVAILABLE` after receiving the first result? That should lead to the following behavior:
1. `ExecuteStreamingSql` is retried, but __without__ a `transaction.begin` and __with__ a `transaction.id` (and with a resume token)
2. There's also no extra `BeginTransaction` RPC

File: test/spanner.ts [hunk 3]
```diff
@@ -2890,6 +2927,30 @@ describe('Spanner with mock server', () => {
...
+          ]).then(([rows]) => {
+            let count = 0;
+            rows.forEach(() => count++);
+            return transaction.commit().then(() => count);
+          });
+        }
+      );
+      assert.strictEqual(rowCount, 3);
+      assert.strictEqual(attempts, 2);
+      await database.close();
```
Comment (olavloite) [inline test/spanner.ts:2951]: Can we add verifications for the requests that we receive here? My expectation would be the following list of requests:

1. `ExecuteStreamingSql` (with `transaction.begin`)
2. `BeginTransaction`
3. `ExecuteStreamingSql` with transaction.id
4. `ExecuteStreamingSql` with transaction.id
5. `Commit`

File: test/spanner.ts [hunk 4]
```diff
@@ -2890,6 +2927,30 @@ describe('Spanner with mock server', () => {
...
+            transaction!.run(selectSql),
+            transaction!.run(selectSql),
+          ]).then(([rows]) => {
+            let count = 0;
+            rows.forEach(() => count++);
+            return transaction.commit().then(() => count);
+          });
+        }
+      );
+      assert.strictEqual(rowCount, 3);
```
Comment (olavloite) [inline test/spanner.ts:2949]: Why is the row count 3? Shouldn't that be 2, as each query returns 1 row.

File: test/spanner.ts (hunk 2 above)
Comment (olavloite) [reply test/spanner.ts:1353]: (It could be that we already have such a test, so check for that first)

File: test/spanner.ts [hunk 5]
```diff
@@ -1314,6 +1314,43 @@ describe('Spanner with mock server', () => {
...
+            streamIndex: index,
+          } as MockError;
+          spannerMock.setExecutionTime(
+            spannerMock.executeStreamingSql,
+            SimulatedExecutionTime.ofError(err)
+          );
+          const database = newTestDatabase();
+
+          await database.runTransactionAsync(async tx => {
+            await Promise.all([tx!.run(selectSql), tx!.run(selectSql)]);
```
Comment (olavloite) [inline test/spanner.ts:1330]: Can we verify here that we get the expected (number of) rows?

File: test/spanner.ts [hunk 6]
```diff
@@ -2890,6 +2952,70 @@ describe('Spanner with mock server', () => {
...
+      );
+      assert.strictEqual(rowCount, 6);
+      assert.strictEqual(attempts, 2);
+      const requests = spannerMock
+        .getRequests()
+        .filter(val => (val as v1.ExecuteSqlRequest).sql)
+        .map(req => req as v1.ExecuteSqlRequest);
+
+      assert.strictEqual(requests.length, 6);
+      assert.ok(
```
[the remark this answers is not in the document]
Comment (alkatrivedi) [reply test/spanner.ts]: Done

File: test/spanner.ts [hunk 7]
```diff
@@ -1314,6 +1314,68 @@ describe('Spanner with mock server', () => {
...
+            requests[1].resumeToken,
+            'Resume token is not set for the retried'
+          );
+          const commitRequests = spannerMock
+            .getRequests()
+            .filter(val => (val as v1.CommitRequest).mutations)
+            .map(req => req as v1.CommitRequest);
+          assert.strictEqual(commitRequests.length, 1);
+          assert.deepStrictEqual(
+            requests[1].transaction!.id,
```
[the remark this answers is not in the document]
Comment (alkatrivedi) [reply test/spanner.ts]: Updated

File: test/spanner.ts (hunk 5 above)
Comment (alkatrivedi) [reply test/spanner.ts]: Done

File: test/spanner.ts (hunk 4 above)
Comment (alkatrivedi) [reply test/spanner.ts]: Updated

File: src/transaction.ts [hunk 8]
```diff
@@ -1219,8 +1227,16 @@ export class Snapshot extends EventEmitter {
           this._update(response.metadata!.transaction);
         }
       })
-      .on('error', () => {
-        if (!this.id && this._useInRunner) {
+      .on('error', err => {
```
Comment (olavloite) [inline src/transaction.ts:1230]: For a different refactor PR: It seems that `runStream` and `createReadStream` have a lot of duplicated code. Could we combine more of those?

File: test/spanner.ts [hunk 9]
```diff
@@ -1314,6 +1314,43 @@ describe('Spanner with mock server', () => {
...
+            'inline txn is not set.'
+          );
+          assert.ok(
+            requests[1].transaction!.id,
+            'Transaction ID is not used for retries.'
+          );
+          assert.ok(
+            requests[1].resumeToken,
+            'Resume token is not set for the retried'
+          );
```
[the remark this answers is not in the document]
Comment (olavloite) [reply test/spanner.ts:1356]: Ack

File: test/spanner.ts (hunk 2 above)
Comment (olavloite) [reply test/spanner.ts:1378]: Ack

File: test/spanner.ts [hunk 10]
```diff
@@ -3211,6 +3326,19 @@ describe('Spanner with mock server', () => {
       assert.ok(!beginTxnRequest, 'beginTransaction was called');
     });
 
+    it('should catch an exception error during invalid queries while using inline begin transaction', async () => {
+      const database = newTestDatabase();
+      await database.runTransactionAsync(async tx => {
+        try {
+          await Promise.all([tx!.run(selectSql), tx!.run(invalidSql)]);
```
Comment (olavloite) [inline test/spanner.ts:3333]: Do we already have a test for the same with these requests in the other order? That is; with the invalid SQL first?

If not, can we add that as well, and verify that:
1. The initial call to ExecuteStreamingSql fails, which also means that there is no transaction ID that is returned.
2. The transaction is retried with an explicit BeginTransaction and then the invalid SQL + working SQL statement.
3. Catch the error from the invalid SQL statement.
4. Commit the transaction (even though one of the statements failed).
5. The result should be (I think):

```
ExecuteSql(transaction{begin: {readwrite=true}})
BeginTransaction(readwrite=true)
ExecuteSql(transaction{id:...})
ExecuteSql(transaction{id:...})
Commit
```

**Review by olavloite — APPROVED** [review]

File: src/transaction.ts (hunk 8 above)
Comment (surbhigarg92) [reply src/transaction.ts:1230]: Sure.

File: test/spanner.ts (hunk 10 above)
Comment (surbhigarg92) [reply test/spanner.ts:3333]: Will be doing this in a separate PR

**Pull request merged**